Add TriviaQA task

SanyamSatia opened the pull request:
Add TriviaQA task with the RC version of the dataset: http://nlp.cs.washington.edu/triviaqa/

alexholdenmiller commented on parlai/tasks/triviaqa/agents.py:
```diff
@@ -0,0 +1,57 @@
+                                                  evidence_filename)
+                with open(evidence_file_path) as evidence_file:
+                    evidence = evidence_file.read()
+                    yield (evidence + '\n' + question, answers), True
```
hm I'm going to talk to Jason about how we want to set up the evidence, I think selecting one random one (fixed at launch time) isn't how we'll want to do this

alexholdenmiller commented on parlai/tasks/triviaqa/agents.py:
```diff
@@ -0,0 +1,57 @@
+
+        for dataset in datasets:
+            dataset_path = os.path.join(path,
+                                        dataset + '-' + self.suffix + '.json')
```
I may want to split these into different teachers (ie. a WikiTeacher and a WebTeacher)

SanyamSatia replied:
Split into different teachers: WikipediaTeacher and WebTeacher. The DefaultTeacher now inherits from MultiTaskTeacher.

SanyamSatia commented:
* Split existing teacher into different teachers and changed DefaultTeacher to inherit from MultiTaskTeacher in order to allow training on both datasets simultaneously. 
* Added download test

alexholdenmiller commented on parlai/tasks/triviaqa/agents.py:
```diff
@@ -0,0 +1,83 @@
+    def setup_data(self, path):
+        return setup_data_common('wikipedia', path, self.suffix,
+                                 self.evidence_dir)
+
```
can you also add a "verified teacher" for the verified data? it will just always use the dev datatype (say, print a warning but don't error if datatype != 'valid'). this won't affect the default teacher

SanyamSatia replied:
Added a VerifiedWikipediaTeacher and VerifiedWebTeacher inheriting from WikipediaTeacher and WebTeacher, respectively. Also added a VerifiedTeacher inheriting from MultiTaskTeacher.

alexholdenmiller commented on parlai/tasks/triviaqa/agents.py:
```diff
@@ -0,0 +1,83 @@
+        if len(evidence_list) == 0:
+            continue
+
+        evidence_num = random.randrange(len(evidence_list))
```
adjust this as we discussed on messenger, thanks! (separate ex for each web evidence, concat all wiki evidence)

SanyamSatia commented:
Changed evidence handling as discussed and added verified teachers

alexholdenmiller commented on parlai/tasks/triviaqa/agents.py:
```diff
@@ -0,0 +1,135 @@
+        dataset = 'web'
+        dataset_path = os.path.join(path,
+                                    self.prefix + dataset + '-' + self.suffix +
+                                    '.json')
```
actually, can you move this to the init, so that `opt['datafile']` points to this? then the user can inspect the exact datafile being loaded by looking at the opt file

SanyamSatia replied:
Moved the initialization of the datafile option to the constructor.

alexholdenmiller commented:
Awesome, thank you!!
